src/coreclr/jit/valuenum.cpp

```diff
@@ -4421,13 +4421,30 @@ bool ValueNumStore::VNEvalCanFoldBinaryFunc(var_types type, VNFunc func, ValueNu
             case GT_RSZ:
             case GT_ROL:
             case GT_ROR:
+               if (IsVNHandle(arg0VN) || IsVNHandle(arg1VN))
```

Comment:
It's unnecessarily conservative to give up on folding handles in the Jit case. I think this should rather refactor `GenTreeIntConCommon::ImmedValCanBeFolded` a little so that it is callable from VN.

Reply:
Sounds reasonable but I am not comfortable enough with the code to do it myself.